Cmorize wfde5

mwjury opened the pull request:
I added a cmorizer for WFDE5 (bias-corrected reconstruction of ERA5). Depends on  #1990.

- [ ] [☝ Create an issue](https://github.com/ESMValGroup/ESMValTool/issues) to discuss what you are going to do

## Checklist

It is the responsibility of the author to make sure the PR is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

- [x] [🛠][1] PR has a descriptive title for the [changelog](https://docs.esmvaltool.org/en/latest/community/introduction.html#branches-pull-requests-and-code-review)
- [x] [🛠][1] Code follows the [style guide](https://docs.esmvaltool.org/en/latest/community/introduction.html#code-style)
- [x] [🛠][1] [Circle/CI tests pass](https://docs.esmvaltool.org/en/latest/community/introduction.html#branches-pull-requests-and-code-review)
- [x] [🛠][1] [Codacy code quality checks pass](https://docs.esmvaltool.org/en/latest/community/introduction.html#branches-pull-requests-and-code-review)

### New or updated [data reformatting script](https://docs.esmvaltool.org/en/latest/develop/dataset.html):

- [x] [🛠][1] Dataset is added to the [table in the documentation](https://docs.esmvaltool.org/en/latest/community/dataset.html#dataset-documentation)
- [x] [🛠][1] Documentation contains [instructions to obtain the data](https://docs.esmvaltool.org/en/latest/community/dataset.html#dataset-documentation)
- [x] [🛠][1] [Tests for the CMORized data](https://docs.esmvaltool.org/en/latest/community/dataset.html#dataset-tests) are available
- [x] [🧪][2] Numbers/units of the data look [physically meaningful](https://docs.esmvaltool.org/en/latest/community/dataset.html#cmorizer-output)
- [ ] [🛠][1] Data set is added to the [OBS data pool](https://docs.esmvaltool.org/en/latest/community/dataset.html#adding-your-dataset-to-the-obs-data-pool) @remi-kazeroni

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

valeriupredoi commented:
this looks ready, I added the `release` tag and we need to fix the iris documentation as per https://github.com/ESMValGroup/ESMValTool/issues/2011

valeriupredoi reviewed (approved):
cheers @mwjury - I will merge once we merge the fix to the iris documentation from #2012 :beer:

mwjury commented:
Depends on #1990 (as written at the top)

bouweandela commented:
Just applying some new labels to make the status of all open pull requests more clear. Note that the tests on CircleCI and the documentation build must be successful (green) before the label 'approved by technical reviewer' can be applied.

remi-kazeroni reviewed (commented):
The version of the data mentioned in this PR (v1.0) is now deprecated. @mwjury coud you please update the downloading instrustions, the version number (v1.1) accordlingly and check whether your script runs fine on the newest data. Thanks!

valeriupredoi commented:
@remi-kazeroni cheers, mate! Next time please use the review changes button and ask for changes so that the official review is logged and the merge can not happen unless both you as reviewer and the PR creator agree on changes made/not needed :+1:

valeriupredoi reviewed (changes requested):
Technical review passed by me, but changes needed for the sci review from [comment](https://github.com/ESMValGroup/ESMValTool/pull/1991#pullrequestreview-587367000)

lukruh commented on esmvaltool/cmorizers/obs/cmor_config/WFDE5.yml:
```diff
@@ -0,0 +1,47 @@
+    add_day: True
+    add_mon: True
+  pr_CRU-GPCC:
+    short_name: pr
```
afaik total precipitation (pr) is the sum of rainfall and snowfall. snowfall is available in WFDE5 as well. Wouldn't it make sense to sum them up?

mwjury replied:
Thanks for pointing this out @lukruh ! I'll change it next week.

hb326 commented:
I think there is a problem with the calculation of the monthly means. I ran the cmorizer and get for the file "OBS_WFDE5_ground_v1.1-CRU+GPCC_Amon_pr_197901-197912.nc" only one time step. I think the monthly means are overwritten when they are stored in the file. Could you check that again?

hb326 commented:
@mwjury, you want to check the calculations of the monthly means for this pull request? I am happy to do the scientific review then.

mwjury commented:
@hb326 In the previous version for v1.0 the output was monthly files (one month per file), so I do not understand how you got files for a year.
I have updated to the current version, could you please have a look.

mwjury commented:
Sorry it took so long. @valeriupredoi this should be ready to go (FLAKE8-check failing due to regex, any ideas).
- [x] Updated to current version (v1.1) @remi-kazeroni .
- [x] Accounted for snowfall as raised by @lukruh .
- [x] Added yearly treatment of files as implemented in cmorize_WFDE5_monthly_fix by @schlunma , removed cmorization of hourly data, but still keeping cmorizing daily data.

hb326 commented on esmvaltool/cmorizers/obs/cmorize_obs_wfde5.py:
```diff
@@ -0,0 +1,135 @@
+import iris
+from cf_units import Unit
+from esmvalcore.preprocessor import daily_statistics, monthly_statistics
+
```
When I pulled the branch to test it yesterday, this line was only showing up as 
`from esmvalcore.preprocessor import monthly_statistics`

That made the cmorizer crash, of course. I added the `daily_statistics` manually, and then it ran fine. 
This is weird. @valeriupredoi, could you check what you get when you pull the branch?

valeriupredoi replied:
it's possible you pulled a local copy, or that your git cache may be a bit special :grin: Anyway I pulled the branch from remote and that was there all fine, I now merged `master` in so all should be tiptop. I'll review it now :+1:

hb326 reviewed (commented):
For me the cmorizer looks good now. I think, providing the option to get "pr" out as the sum from rainfall flux and snowfall flux is great! I also get 12 (different) months now in the monthly file, and 365 (different) days for the daily files.

zklaus commented:
> (FLAKE8-check failing due to regex, any ideas).

Yes! According to https://www.flake8rules.com/rules/W605.html you should write your regex in a raw string, i.e. `r'bla'`. The reason is that without the `r`, Python will replace all occurrences of a backslash followed by something with a control character right then and there, before even passing it to the regex library. So `'\n'` becomes a newline, `'\r'` a carriage return, etc. The reason why it still works as it stands is that the backslashes are not followed by any recognized control sequences and therefore are in the end not replaced. This is not completely accidental since back in the days the designers of regular expressions had this in mind, but it might also not be guaranteed, which is why Flake8 is saying it's better to tell Python that you don't want it replacing backslashes here.

valeriupredoi commented:
> > (FLAKE8-check failing due to regex, any ideas).
> 
> Yes! According to https://www.flake8rules.com/rules/W605.html you should write your regex in a raw string, i.e. `r'bla'`. The reason is that without the `r`, Python will replace all occurrences of a backslash followed by something with a control character right then and there, before even passing it to the regex library. So `'\n'` becomes a newline, `'\r'` a carriage return, etc. The reason why it still works as it stands is that the backslashes are not followed by any recognized control sequences and therefore are in the end not replaced. This is not completely accidental since back in the days the designers of regular expressions had this in mind, but it might also not be guaranteed, which is why Flake8 is saying it's better to tell Python that you don't want it replacing backslashes here.

`# noqa` FTW

valeriupredoi reviewed (changes requested):
good stuff @hb326 - couple minor comments from me :beer:

mwjury commented on esmvaltool/cmorizers/obs/cmorize_obs_wfde5.py:
```diff
@@ -0,0 +1,135 @@
+
+    # time points are XX:00:00, should be XX:30:00
+    time = cube.coord('time')
+    time.points = time.points + 1 / 48
```
Hm, I added this to have the points centred for an hour. I am not sure which is better / correct tbh.
If I don't do it, subsequent bounds are:
-  [1978-12-31 23:30:00, 1979-01-01 23:30:00]  for daily data
-  [1978-12-31 23:30:00, 1979-01-31 23:30:00]  for monthly data

mwjury commented on esmvaltool/cmorizers/obs/cmorize_obs_wfde5.py:
```diff
@@ -0,0 +1,135 @@
+    version = cfg['attributes']['version'] + '-' + var['reference']
+
+    cubes = iris.load(filenames)
+    cube = cubes.concatenate_cube()
```
That's iris internal. If it fails it should return a list instead of a single cube which will lead to errors down the road.

remi-kazeroni commented:
I have just run this cmorizer to put the cmorized data in the obs folder. I realized there is a typo in the `WFDE5.yml` config file which should have: `tier: 2` and not `tier: 3`. The cmorizer runs fine but the `recipe_check_obs.yml` not... Sorry, I should have jumped in here earlier and see this while cmorizing the data before this is merged. But in such cases @valeriupredoi please ping me before merging and I'll do my best to have a look asap 👍 The last box was not checked and we agreed that this is my task 😄 Shall I restore the branch or fix this issue in another PR?

valeriupredoi commented:
ouch, totally my bad, sorry - I did notice the last box being unchecked but I thought someone forgot to check it :grin: No need to restore the branch, just do a quick PR with the typo correction and I'll approve it pronto :+1:
